memas/corpus/corpus_searching.py

```diff
@@ -1,17 +1,55 @@
 # from search_redirect import SearchSettings
 from uuid import UUID
 from functools import reduce
-from memas.interface.corpus import Corpus, CorpusFactory
+from memas.interface.corpus import Corpus, CorpusFactory, CorpusType
 from memas.interface.corpus import Citation
+from collections import defaultdict 
 from memas.interface.storage_driver import DocumentEntity
 from memas.interface.exceptions import SentenceLengthOverflowException
 
 
-def corpora_search(corpus_ids: list[UUID], clue: str) -> list[tuple[float, str, Citation]]:
+def mult_corpus_search(corpus_sets : dict[Corpus], clue, ctx, result_limit) -> list[tuple[float, str, Citation]]:
+    results = defaultdict(list)
+
+    # Direct each multicorpus search to the right algorithm
+    for corpusType, corpora_list in corpus_sets.items() :
+        # Default basic corpus handling
+        if corpusType == CorpusType.KNOWLEDGE or corpusType == CorpusType.CONVERSATION :
+            corpus_type_results = basic_corpora_search(corpora_list, clue, ctx)
+            results["BASIC_SCORING"].extend(corpus_type_results)
+
+
+    sorted_results_matrix = []
+    # Sort results with compareable scoring schemes
+    for scored_results in results.values() :
+        # Sort by descending scoring so best results come first
+        sorted_scored_results = sorted(scored_results, key=lambda x: x[0], reverse=True)
+        sorted_results_matrix.append(sorted_scored_results)
+
+    # To combine results for corpora that don't have compareable scoring take equal sized subsets of each Corpus type
+    # TODO : Consider changing this at some point in the future to have better searching of corpus sets with non-comparable scoring
+    combined_results = []
+    for j in range(max([len(x) for x in sorted_results_matrix])) :
+        for i in range(len(sorted_results_matrix)) :
+            if j >= len(sorted_results_matrix[i]) or len(combined_results) >= result_limit:
+                break
+            combined_results.append(sorted_results_matrix[i][j])
+        if len(combined_results) >= result_limit:
+            break
+
+    return combined_results
+
+"""
+All corpora here should be of the same CorpusType implementation (basic_corpus)
+"""
+def basic_corpora_search(corpora: list[Corpus], clue: str, ctx) -> list[tuple[float, str, Citation]]:
```

Comment:
this entire function looks pretty sus... It looks like we're jumping out of the corpus implementation and redoing the basic corpus search again? **We can keep this for now**, but it'd be best to implement this in a way that properly modularizes the logic (feel free to refactor any interfaces that get in the way).

Reply:
You're right that it is basically a redo of basic corpus search, which isn't ideal. Of course modular code is the way to go, and I was thinking the same thing when I worked on that function. It isn't straightforward without a larger refactor, which I didn't want to do before talking to you. I can work on a possible refactor for that and then we can discuss it.

Reply:
Ah gotcha, then sure let's keep this for now